internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala

```diff
@@ -147,142 +245,192 @@ object Terminal {
+    private[this] val runnable: Runnable = () => {
+      @tailrec def impl(): Unit = {
+        val b = in.read
+        buffer.put(b)
+        if (b != -1) impl()
+        else closed.set(true)
+      }
       try {
-        flushQueue.take()
-        val bytes = new java.util.ArrayList[Byte]
-        writeLock.synchronized {
-          lineBuffer.drainTo(bytes)
-          import scala.collection.JavaConverters._
-          val remaining = bytes.asScala.foldLeft(new ArrayBuffer[Byte]) { (buf, i) =>
-            if (i == 10) {
-              ProgressState.addBytes(buf)
-              ProgressState.clearBytes()
-              buf.foreach(b => originalOut.write(b & 0xFF))
-              ProgressState.reprint(originalOut)
-              currentLine.set(new ArrayBuffer[Byte])
-              new ArrayBuffer[Byte]
-            } else buf += i
-          }
-          if (remaining.nonEmpty) {
-            currentLine.get ++= remaining
-            originalOut.write(remaining.toArray)
-          }
-          originalOut.flush()
-        }
-      } catch { case _: InterruptedException => isStopped.set(true) }
-      if (!isStopped.get) run()
+        latch.await()
+        impl()
+      } catch { case _: InterruptedException => }
+    }
+    executor.submit(runnable)
+    override def read(): Int = {
+      latch.countDown()
+      takeOne match {
+        case -1 => throw new ClosedChannelException
+        case b  => b
+      }
+    }
+
+    override def available(): Int = {
+      latch.countDown()
+      buffer.size
+    }
+    override def close(): Unit = {
+      executor.shutdownNow()
+      ()
     }
   }
+  private[this] val nonBlockingIn: WriteableInputStream =
+    new WriteableInputStream(jline.TerminalFactory.get.wrapInIfNeeded(originalIn), "console")
+  private[this] val inputStream = new AtomicReference[InputStream](System.in)
   private[this] def withOut[T](f: => T): T = {
-    val thread = new WriteThread
     try {
-      System.setOut(SystemPrintStream)
-      scala.Console.withOut(SystemPrintStream)(f)
+      System.setOut(proxyPrintStream)
+      scala.Console.withOut(proxyOutputStream)(f)
     } finally {
-      thread.close()
       System.setOut(originalOut)
     }
   }
   private[this] def withIn[T](f: => T): T =
     try {
-      System.setIn(Terminal.wrappedSystemIn)
-      scala.Console.withIn(Terminal.wrappedSystemIn)(f)
+      inputStream.set(Terminal.wrappedSystemIn)
+      System.setIn(wrappedSystemIn)
+      scala.Console.withIn(proxyInputStream)(f)
     } finally System.setIn(originalIn)
 
-  private[sbt] def withPrintStream[T](f: PrintStream => T): T = writeLock.synchronized {
-    f(originalOut)
+  private[sbt] def withPrintStream[T](f: PrintStream => T): T = console.withPrintStream(f)
+  private[this] val attached = new AtomicBoolean(true)
+  private[this] val terminalHolder = new AtomicReference(wrap(jline.TerminalFactory.get))
+  private[this] val currentTerminal = new AtomicReference[Terminal](terminalHolder.get)
```

Comment:
These two (terminalHolder and currentTerminal) don't seem to be well managed at the moment.  Or, at the very least, not obviously so.

To start with `jline.TerminalFactory.get` -> `terminalHolder` -> `currentTerminal` (and then immediately `jline.TerminalFactory` is change to be the JLine1/JLine2 wrapper.)

But then `terminalHolder` is read from in `console` and re-initialised in `reset`, while `currentTerminal` is read from in all the proxy components (`proxyInputStream`, `proxyOutputStream`, `ProxyTerminal` and `get`) and changed in `set`.

Is all that correct?  I'm guessing `terminalHolder` is meant to be "live" and `currentTerminal` is meant to be a more persistent instance that proxies for whatever the terminal is actually in use?

A couple of paragraphs on the ideas in play and how to use it (as the `Terminal` object's Scaladoc) would come very handy! ❤️

Also, seeing as there's all 4 adapters/forwarders for the terminals:

* "wrap": jline1 -> jline1/jline2
* "toJLine": Terminal -> jline1/jline2
* "ProxyTerminal": Terminal -> Terminal
* "ConsoleTerminal": jline1/jline2 -> Terminal

I was almost thinking it might be easier to understand and keep right if the terminal stream bits were in a separate object/file.  Not because 600 lines is too much, but because it's complicated by the mutation going around.

Reply:
I'll look into this more tomorrow. I'm not sure how much can be done about the adapters without internally passing around instances of `jline.Terminal2`, which I'm not keen to do. That being said, there's no doubt the organization could be better. The question is how much effort I really want to put into reorganizing. It's at a point where I have spent an absurd amount of time on this project so I'm running out of steam for refactoring.

Reply:
I can propose those changes, and that would work in my favour as it avoids me rebasing/rewriting my local state.

But wdyt about the state of the terminal holder/currentTerminal setup?

Reply:
After performing a few days of git surgery on this PR, I was really unexcited about further changes last night. I'm feeling more sanguine today. I renamed some fields and added some commentary in https://github.com/sbt/sbt/pull/5620/commits/5238c78dfd12d0fac4a5ec588be0d069eaf059b7